crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -1655,7 +1839,9 @@ fn guest_read<T>(
                     BLOCKED
                 }
 
-                WriteState::Closed => CLOSED,
+                // If at some point the writer chose to close, the final stream that comes back
+                // should contain CLOSED and the error context
+                WriteState::Closed(err_ctx) => CLOSED | err_ctx as usize,
```

Comment:
We need to lower this, `ErrorContext::lower_to_index`-style rather than return it directly.